hydroflow_lang/src/graph/ops/lattice_join.rs

```diff
@@ -0,0 +1,206 @@
+///
+/// You must specify the LHS and RHS lattice types, they cannot be inferred.
+///
+/// ```hydroflow
+/// // should print `(key, (2, 1))`
+/// my_join = lattice_join::<hydroflow::lang::lattice::ord::MaxRepr<usize>, hydroflow::lang::lattice::ord::MaxRepr<usize>>();
+/// source_iter(vec![("key", 0), ("key", 2)]) -> [0]my_join;
+/// source_iter(vec![("key", 1)]) -> [1]my_join;
+/// my_join -> for_each(|(k, (v1, v2))| println!("({}, ({}, {}))", k, v1, v2));
+/// ```
+///
+/// `lattice_join` can also be provided with one or two generic lifetime persistence arguments, either
+/// `'tick` or `'static`, to specify how join data persists. With `'tick`, pairs will only be
+/// joined with corresponding pairs within the same tick. With `'static`, pairs will be remembered
+/// across ticks and will be joined with pairs arriving in later ticks. When not explicitly
+/// specified persistence defaults to `static.
+///
+/// When two persistence arguments are supplied the first maps to port `0` and the second maps to
+/// port `1`.
+/// When a single persistence argument is supplied, it is applied to both input ports.
+/// When no persistence arguments are applied it defaults to `'static` for both.
+///
+/// The syntax is as follows:
+/// ```hydroflow,ignore
+/// lattice_join::<MaxRepr<usize>, MaxRepr<usize>>(); // Or
+/// lattice_join::<'static, MaxRepr<usize>, MaxRepr<usize>>();
+///
+/// lattice_join::<'tick, MaxRepr<usize>, MaxRepr<usize>>();
+///
+/// lattice_join::<'static, 'tick, MaxRepr<usize>, MaxRepr<usize>>();
+///
+/// lattice_join::<'tick, 'static, MaxRepr<usize>, MaxRepr<usize>>();
+/// // etc.
+/// ```
+///
+/// ### Examples
+///
+/// ```rustbook
+/// async fn test_lattice_join() {
+///     use hydroflow::lang::lattice::ord::MaxRepr;
+///
+///     let (input_send, input_recv) = hydroflow::util::unbounded_channel::<(usize, usize)>();
+///     let (out_tx, mut out_rx) = hydroflow::util::unbounded_channel::<(usize, (usize, usize))>();
+///
+///     let mut df = hydroflow::hydroflow_syntax! {
+///         my_join = lattice_join::<'tick, MaxRepr<usize>, MaxRepr<usize>>();
+///         source_iter([(7, 2), (7, 1)]) -> [0]my_join;
+///         source_stream(input_recv) -> [1]my_join;
+///         my_join -> for_each(|v| out_tx.send(v).unwrap());
+///     };
+///     input_send.send((7, 5)).unwrap();
+///     df.run_tick();
+///     let out: Vec<_> = hydroflow::util::collect_ready(&mut out_rx);
+///     assert_eq!(out, vec![(7, (2, 5))]);
+/// }
+/// ```
+#[hydroflow_internalmacro::operator_docgen]
+pub const LATTICE_JOIN: OperatorConstraints = OperatorConstraints {
+    name: "lattice_join",
+    hard_range_inn: &(2..=2),
+    soft_range_inn: &(2..=2),
+    hard_range_out: RANGE_1,
+    soft_range_out: RANGE_1,
+    num_args: 0,
+    persistence_args: &(0..=2),
+    type_args: &(2..=2),
+    is_external_input: false,
+    ports_inn: Some(|| super::PortListSpec::Fixed(parse_quote! { 0, 1 })),
+    ports_out: None,
+    properties: FlowProperties {
```

Comment:
looking for feedback on how this operator interacts with the flow properties, I'm not sure if it is monotonic/deterministic/etc.

Reply:
I think `preserve, preserve, false` is right

Reply:
We might need a more complex mechanism for multi-input operators. but for now the flowproperties are unused